lib/src/fsmonitor.rs

```diff
@@ -0,0 +1,208 @@
+//! virtualized filesystem.
+
+#![warn(missing_docs)]
+
+use std::path::PathBuf;
+use std::str::FromStr;
+
+/// The recognized kinds of filesystem monitors.
+pub enum FsmonitorKind {
+    /// The Watchman filesystem monitor (https://facebook.github.io/watchman/).
+    Watchman,
+
+    /// Only used in tests.
+    Test {
+        /// The set of changed files to pretend that the filesystem monitor is
+        /// reporting.
+        changed_files: Vec<PathBuf>,
+    },
+}
+
+impl FromStr for FsmonitorKind {
+    type Err = config::ConfigError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "watchman" => Ok(Self::Watchman),
+            "test" => Err(config::ConfigError::Message(
+                "cannot use test fsmonitor in real repository".to_string(),
+            )),
+            other => Err(config::ConfigError::Message(format!(
+                "unknown fsmonitor kind: {}",
+                other
+            ))),
+        }
+    }
+}
+
+/// Filesystem monitor integration using Watchman
+/// (https://facebook.github.io/watchman/). Requires `watchman` to already be
+/// installed on the system.
+#[cfg(feature = "watchman")]
+pub mod watchman {
+    use std::path::{Path, PathBuf};
+    use std::str::FromStr;
+
+    use itertools::Itertools;
+    use thiserror::Error;
+    use tracing::info;
+    use watchman_client::prelude::{
+        Clock as InnerClock, ClockSpec, NameOnly, QueryRequestCommon, QueryResult,
+    };
+
+    /// Represents an instance in time from the perspective of the filesystem
+    /// monitor.
+    ///
+    /// This can be used to perform incremental queries. When making a query,
+    /// the result will include an associated "clock" representing the time
+    /// that the query was made.  By passing the same clock into a future
+    /// query, we inform the filesystem monitor that we only wish to get
+    /// changed files since the previous point in time.
+    #[derive(Clone, Debug)]
+    pub struct Clock(InnerClock);
+
+    impl ToString for Clock {
+        fn to_string(&self) -> String {
+            let Self(clock) = self;
+            match clock {
+                InnerClock::Spec(spec) => match spec {
+                    ClockSpec::StringClock(s) => s.clone(),
+                    ClockSpec::UnixTimestamp(i) => i.to_string(),
+                },
+                InnerClock::ScmAware(_) => {
+                    unimplemented!("SCM-aware Watchman clocks not supported")
+                }
+            }
+        }
+    }
+
+    impl FromStr for Clock {
+        type Err = <i64 as FromStr>::Err;
+
+        fn from_str(s: &str) -> Result<Self, Self::Err> {
+            let spec = if s.starts_with("c:") {
+                ClockSpec::StringClock(s.to_owned())
+            } else {
+                let timestamp: i64 = s.parse()?;
+                ClockSpec::UnixTimestamp(timestamp)
+            };
+            Ok(Clock(InnerClock::Spec(spec)))
+        }
+    }
```

Comment:
What do you think about giving `Clock` a more direct representation in the `.proto` file? Instead of `string watchman_clock = 4;`, you would do something like this:

```
oneof watchman_clock {
  string string_clock = 4;
  int64 unix_timestamp = 5;
}
```

That would be easier to extend if there's ever another type of `ClockSpec` we want to support.

Reply:
Decided to use a `oneof` in a submessage.